src/irmin-pack/unix/io.ml

```diff
@@ -240,4 +240,101 @@ module Unix = struct
       Sys.remove path;
       Ok ()
     with Sys_error msg -> Error (`Sys_error msg)
+
+  (* Async using fork/waitpid*)
+
+  module Exit = struct
+    let proc_list = ref []
+    let m = Mutex.create ()
+
+    let add gc =
+      Mutex.lock m;
+      proc_list := gc :: !proc_list;
+      Mutex.unlock m
+
+    let remove gc =
+      Mutex.lock m;
+      proc_list := List.filter (fun gc' -> gc <> gc') !proc_list;
+      Mutex.unlock m
+
+    let clean_up () =
+      List.iter
+        (fun gc ->
+          try Unix.kill gc 9
+          with Unix.Unix_error (e, s1, s2) ->
+            [%log.warn
+              "Killing gc process with pid %d failed with error (%s, %s, %s)" gc
+                (Unix.error_message e) s1 s2])
+        !proc_list
+  end
+
+  (* Register function to be called when process terminates. If there is a gc
+     process running, make sure to terminate it. *)
+  let () = at_exit Exit.clean_up
+
+  type status = [ `Running | `Success | `Cancelled | `Failure of string ]
+  type task = { pid : int; mutable status : status }
+
+  let async f =
+    Stdlib.flush_all ();
+    match Lwt_unix.fork () with
+    | 0 ->
+        Lwt_main.Exit_hooks.remove_all ();
+        Lwt_main.abandon_yielded_and_paused ();
+        f ();
+        (* Once the gc is finished, the child process kills itself to
+           avoid calling at_exit functions in upstream code. *)
+        Unix.kill (Unix.getpid ()) 9;
+        assert false (* unreachable *)
+    | pid ->
+        Exit.add pid;
+        { pid; status = `Running }
+
+  let status_of_process_status = function
+    | Lwt_unix.WSIGNALED -7 ->
+        `Success (* the child is killing itself when it's done *)
+    | Lwt_unix.WSIGNALED n -> `Failure (Fmt.str "Signaled %d" n)
+    | Lwt_unix.WEXITED n -> `Failure (Fmt.str "Exited %d" n)
+    | Lwt_unix.WSTOPPED n -> `Failure (Fmt.str "Stopped %d" n)
+
+  let cancel t =
+    let () =
+      match t.status with
+      | `Running ->
+          let pid, _ = Unix.waitpid [ Unix.WNOHANG ] t.pid in
+          (* Do not block if no child has died yet. In this case the waitpid
+             returns immediately with a pid equal to 0. *)
+          if pid = 0 then (
+            Unix.kill t.pid 9;
+            Exit.remove t.pid)
+      | _ -> ()
+    in
+    t.status <- `Cancelled
+
+  let status t : status =
+    match t.status with
+    | `Running ->
+        let pid, status = Unix.waitpid [ Unix.WNOHANG ] t.pid in
+        (* Do not block if no child has died yet. In this case the waitpid
+           returns immediately with a pid equal to 0. *)
+        if pid = 0 then (
+          Unix.kill t.pid 9;
```

Comment:
Am I misreading this? My understanding is `pid = 0` means the task process hasn't died yet. It doesn't seem right that checking the status of a task would kill its process.